substrate-examples/offchain-worker/README.md

```diff
@@ -0,0 +1,70 @@
+# Proof of Concept: Tag Analysis on Origin
+
+This is a copy of [substrate's pallet template](https://github.com/substrate-developer-hub/substrate-node-template/tree/e0c480c0f322d0b0d1b310c93fa646fc0cfdd2df/pallets/template), enriched with a MIRAI [tag-analysis](https://github.com/facebookexperimental/MIRAI/blob/main/documentation/TagAnalysis.md). It is a first proof-of-concept for a [validation of unsigned transactions](https://github.com/bhargavbh/MIRAI/blob/main/substrate_examples/unsigned-transaction/description.md).
+
+# Running
+
+To run the analysis [install mirai](https://github.com/facebookexperimental/MIRAI/blob/main/documentation/InstallationGuide.md) and run
+
+`cargo mirai`
+
+from within this folder. The [config.toml](.cargo/config.toml) makes sure, that the analysis only runs on the function [`mirai_check.code_to_analyze`](src/mirai.rs).
+
+# Tag Analysis
+We use [tag analysis](https://github.com/facebookexperimental/MIRAI/blob/main/documentation/TagAnalysis.md) from MIRAI. In this example, we want to verify that all parameters are properly verified before we call `transform_data`. The tags are set in `validate_transaction_parameters` once the parameters pass the validation checks:
+``` rust
+    let current_block = <system::Pallet<T>>::block_number();
+    if &current_block < block_number {
+        return InvalidTransaction::Future.into()
+    }
+    add_tag!(block_number, ParameterVerified);
+    
+    // Arbitrary check that price is bigger than some value
+    if 100 > *new_price {
+        return InvalidTransaction::Future.into()
+    }
+    add_tag!(new_price, ParameterVerified);
+```
+
+The `transform_data` function is currently not doing anything but it requires that the two parameters have the `ParameterVerified` tag:
+
+``` rust
+    #[requires(has_tag!(new_price, ParameterVerified))]
+    #[requires(has_tag!(block_number, ParameterVerified))]
+    fn transform_data(
+        block_number: &T::BlockNumber,
+        new_price: &u32,
+    ) {
+
+    }
+```
+
+## Output
+
+There are two calls to `validate_transaction_parameters` in `validate_unsigned`. 
+In one case `transform_data` is called without considering the return value of `validate_transaction_parameters` and because of that MIRAI is raising an issue about an unsatisfied precondition:
```

Comment:
I would rename the function "transform_data" to "check_data" or "mirai_check". It's not really doing anything else ...